src/backend/drm/compositor/mod.rs

```diff
@@ -2971,89 +2980,117 @@ where
 
-        let cursor_buffer_size = self.cursor_size.to_logical(1).to_buffer(1, Transform::Normal);
-        let offscreen_buffer = match renderer.create_buffer(DrmFourcc::Argb8888, cursor_buffer_size) {
-            Ok(buffer) => buffer,
-            Err(err) => {
-                debug!(
-                    "failed to create offscreen buffer for cursor {:?}: {}",
-                    plane_info.handle, err
-                );
-                return None;
-            }
-        };
-
-        if let Err(err) = renderer.bind(offscreen_buffer) {
-            debug!(
-                "failed to bind cursor buffer for cursor {:?}: {}",
-                plane_info.handle, err
-            );
+        let Some(storage) = element.underlying_storage(renderer) else {
+            trace!("Can't obtain cursor's underlying storage");
             return None;
         };
 
-        // Try to claim the plane, if this fails we can not use it
-        let plane_claim = match self.surface.claim_plane(plane_info.handle) {
-            Some(claim) => claim,
-            None => {
-                trace!("failed to claim {:?}", plane_info.handle);
-                return None;
-            }
-        };
-
-        // save the renderer debug flags and disable all for the cursor plane
-        let renderer_debug_flags = renderer.debug_flags();
-        renderer.set_debug_flags(DebugFlags::empty());
-
-        let mut render = || {
-            let mut frame = renderer.render(self.cursor_size, output_transform)?;
-
-            frame.clear(
-                [0f32, 0f32, 0f32, 0f32],
-                &[Rectangle::from_loc_and_size((0, 0), self.cursor_size)],
-            )?;
-
-            let src = element.src();
-            let dst = Rectangle::from_loc_and_size((0, 0), element_geometry.size);
-            element.draw(&mut frame, src, dst, &[dst])?;
-
-            frame.finish()?.wait();
+        let cursor_buffer_size = self.cursor_size.to_logical(1).to_buffer(1, Transform::Normal);
 
-            Ok::<(), <R as Renderer>::Error>(())
-        };
+        // Create a pixman image from the source cursor data. This will either be set by the
+        // client, or the compositor's choice.
+        let cursor_texture = match storage {
+            UnderlyingStorage::Wayland(buffer) => self
+                .pixman_renderer
+                .import_buffer(buffer.deref(), None, &[])
+                .transpose()
+                .ok()
+                .flatten(),
+            UnderlyingStorage::Memory((data, format, width, height)) => {
+                let stride =
+                    width * (format::get_bpp(format).expect("Format with unknown bits per pixel") / 8);
+                let size = stride * height;
+
+                let Ok(pixman_format) = pixman::FormatCode::try_from(format) else {
+                    debug!("No pixman format for {format}");
+                    return None;
+                };
+                match pixman::Image::from_slice_mut(
```

Comment:
Instead of `from_slice_mut` with `slice::from_raw_parts_mut` it should be possible to use
`Image::from_raw_mut`